### DISCUSSION TITLE
picolibc usage with ARM SWO (Single Wire Output)

### DISCUSSION THREAD

**Discussion by TomaszOsypinski** [picolibc/picolibc#741] [Q&A] [answered]
Hi everyone,

I would like to use picolib without semihost. 

Instead of semihost I would like to use SWO for printf.

This scheme working nice with Newlib with retargeting function _write.

Any suggestion how to do it in easy way for picolib?

**Comment by keith-packard** [comment, marked as answer]
It's pretty easy to adapt any output mechanism; you replace the implementation of the stdin/stdout/stderr structs; those contain function pointers to put and get a single byte. There are examples all over the tree. 

Down inside the tinystdio implementation itself, you'll find https://github.com/picolibc/picolibc/blob/main/newlib/libc/tinystdio/posixiob_stdout.c and friends which use POSIX functions for this. 

Over in the semihost directory, there are a selection of different implementations for various emulation environments like https://github.com/picolibc/picolibc/blob/main/semihost/iob.c which uses ARM-style semihosting, https://github.com/picolibc/picolibc/blob/main/semihost/machine/x86/e9_io.c which uses x86 "e9" debugging and https://github.com/picolibc/picolibc/blob/main/semihost/machine/sparc/sparc_iob.c which shows how to use a simulated UART for stdout.

Create your replacement versions of stdin, stdout and stderr and then link that into your application ahead of libc so that you select your versions instead of falling back to the C library versions.

You can see a sample project which uses picolibc and replaces these bits here: https://github.com/keith-packard/snek/blob/main/chips/samd21/ao-stdio.c -- that stacks picolibc stdio on top of the USB bits in that hardware.

**Reply by TomaszOsypinski** [reply]
Dear Keith,

Thank you for your fast response. Example was very useful for me.
I would like to show my appreciation for your work. Picolibc is awesome. 

Best regards,
Tom
